### PR TITLE
Feature: Embed total_length() function

### PR DESCRIPTION
**Pull request by zykrahgaming**
### Summary

See #795 . Opened due to renaming branch for consistency.

### Checklist

- [x] I have run `nox` and all the pipelines have passed.
- [x] I have made unittests according to the code I have added/modified/deleted.

### CODE REVIEW

File: hikari/embeds.py [hunk 1]
```diff
@@ -991,3 +991,30 @@ def __eq__(self, other: typing.Any) -> bool:
...
+    def total_length(self) -> int:
+        """Return the total length/character count of this embed.
+
+        Useful for determining if the embed exceeds Discord's 6000 character
+        limit. Accounts for this embed's title, description, fields (all field
+        names and values), footer, and author.
+
+        Returns
+        -------
+        int
```
Comment (davfsa) [inline hikari/embeds.py:1004]:
```suggestion
        builtins.int
```

File: hikari/embeds.py [hunk 2]
```diff
@@ -991,3 +991,30 @@ def __eq__(self, other: typing.Any) -> bool:
             else:
                 return True
         return False
+
+    def total_length(self) -> int:
+        """Return the total length/character count of this embed.
+
+        Useful for determining if the embed exceeds Discord's 6000 character
```
Comment (davfsa) [inline hikari/embeds.py:998]: Do we really want to document what the current maximum is?

Docstring could just be left to:
```
"""Get the total character count of the embed, including fields, author and footer.
""""
```

**Review by davfsa — CHANGES_REQUESTED** [review]

**Review by davfsa — COMMENTED** [review]
Could you please add testing for this?

**Comment by zykrahgaming** [comment]
Test added, waiting for CI.

File: tests/hikari/test_embeds.py [hunk 3]
```diff
@@ -62,3 +62,22 @@ def test_proxy_filename(self, resource_with_proxy):
...
+class TestEmbedTotalLength:
+    @pytest.fixture()
+    def embed(self):
+        embed = embeds.Embed()
+        return embed
+
+    def test_total_length_when_embed_is_empty(self, embed):
+        assert embed.total_length() == 0
+
+    def test_total_length(self, embed):
```
Comment (davfsa) [inline tests/hikari/test_embeds.py:76]: For better test coverage, there should also be cases for when each of them are None or not.

You can check the html files generated by `nox -s pytest` under `public/` for coverage statistics :)

**Review by davfsa — CHANGES_REQUESTED** [review]

File: tests/hikari/test_embeds.py (hunk 3 above)
Comment (zykrahgaming) [reply tests/hikari/test_embeds.py:76]: Ah yea I was going to do that initially but thought it was pedantic. Will add now

File: tests/hikari/test_embeds.py [hunk 4]
```diff
@@ -62,3 +62,22 @@ def test_proxy_filename(self, resource_with_proxy):
     def test_proxy_filename_when_resource_is_none(self, resource_with_proxy):
         resource_with_proxy.proxy_resource = None
         assert resource_with_proxy.proxy_filename is None
+
+
+class TestEmbedTotalLength:
+    @pytest.fixture()
+    def embed(self):
+        embed = embeds.Embed()
+        return embed
```
[the remark this answers is not in the document]
Comment (zykrahgaming) [reply tests/hikari/test_embeds.py]: Removed

File: tests/hikari/test_embeds.py (hunk 3 above)
Comment (zykrahgaming) [reply tests/hikari/test_embeds.py]: Added functions, full coverage now

File: changes/796.feature.md [hunk 5]
```diff
@@ -0,0 +1,3 @@
+Added a total_length() function to hikari.Embed
+- Takes into account the character length of the embed's title, description, fields (all field names and values), footer, and author combined.
+- Useful for determining if the embed exceeds Discord's 6000 character limit.
```
Comment (davfsa) [inline changes/796.feature.md:3]: Again, not worth documenting the character limit

File: changes/796.feature.md [hunk 6]
```diff
@@ -0,0 +1,3 @@
+Added a total_length() function to hikari.Embed
```
Comment (davfsa) [inline changes/796.feature.md:1]:
```suggestion
Added a `total_length` function to `hikari.embeds.Embed`
```

**Review by davfsa — APPROVED** [review]
LGTM, just the small comment

File: hikari/embeds.py [hunk 7]
```diff
@@ -991,3 +991,26 @@ def __eq__(self, other: typing.Any) -> bool:
...
+
+        if self._fields:
+            for field in self._fields:
+                total += len(field.name) + len(field.value)
+
+        if self._footer is not None:
+            total += len(self._footer.text or "")
+
+        if self._author is not None:
+            total += len(self._author.name or "")
```
Comment (FasterSpeeding) [inline hikari/embeds.py:1014]: just a random style question but for this and the footer one could this not just be
```py
if self._author and self._author.name:
    total += len(self._author.name)
```
feel like that might be a bit cleaner

Comment (davfsa) [reply hikari/embeds.py:1014]: I had a try at using `getattr` when this got opened. It's much more efficient than checking it twice, but it makes it a bit awkward to read.

File: changes/796.feature.md (hunk 6 above)
Comment (FasterSpeeding) [reply changes/796.feature.md:1]: This seems a tad bit overly pedantic and i don't think it should block a pr lol

File: changes/796.feature.md (hunk 5 above)
Comment (FasterSpeeding) [reply changes/796.feature.md:3]: Sane here, changelogs are essentially frame so any data referenced in it (or any changes for that matter) shouldn't have to stay up to date anyways

**Comment by FasterSpeeding** [comment]
Btw as a case point for this, this'll be helpful if we ever decide to add validation to the interaction response builders (and the use case for adding validation there specifically is that you don't get any feedback from discord as to why your response was bad)

File: hikari/embeds.py (hunk 7 above)
Comment (zykrahgaming) [reply hikari/embeds.py]: Changed

File: changes/796.feature.md (hunk 5 above)
Comment (zykrahgaming) [reply changes/796.feature.md:3]: Will leave it in that case

**Comment by zykrahgaming** [comment]
> just a random style question but for this and the footer one could this not just be
> 
> ```python
> if self._author and self._author.name:
>     total += len(self._author.name)
> ```
> 
> feel like that might be a bit cleaner

This gives the error 
```
    def total_length(self) -> int:
        """Get the total character count of the embed.
    
        Returns
        -------
        builtins.int
            The total character count of this embed, including title, description,
            fields, footer, and author combined.
        """
        total = len(self._title or "") + len(self._description or "")
    
        if self._fields:
            for field in self._fields:
                total += len(field.name) + len(field.value)
    
>       if self._footer and self._author.name:
E       AttributeError: 'NoneType' object has no attribute 'name'
```

**Comment by zykrahgaming** [comment]
Updated/merged branch to fix nox safety failling linting (aiohttp was outdated).

**Review by FasterSpeeding — APPROVED** [review]

**Pull request merged**